sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala

```
@@ -394,3 +395,149 @@ case class HllUnionAgg(
+  }
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ApproxTopK = {
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+  }
+
+  override protected def withNewChildrenInternal(
+                                                  newLeft: Expression,
+                                                  newRight: Expression): ApproxTopK = {
+    copy(left = newLeft, right = newRight)
+  }
+
+  override def update(sketch: ItemsSketch[Any], input: InternalRow): ItemsSketch[Any] = {
+    val v = left.eval(input)
+    if (v != null) {
+      left.dataType match {
+        case _: IntegerType => sketch.update(v.asInstanceOf[Int])
+        case _: LongType => sketch.update(v.asInstanceOf[Long])
+        case _: DoubleType => sketch.update(v.asInstanceOf[Double])
+        case st: StringType =>
+          val cKey = CollationFactory.getCollationKey(v.asInstanceOf[UTF8String], st.collationId)
+          sketch.update(cKey.toString)
+        case dataType => throw new SparkUnsupportedOperationException(
+          errorClass = "_LEGACY_ERROR_TEMP_3263",
+          messageParameters = Map("dataType" -> dataType.toString))
+      }
+    }
+    sketch
+  }
+
+  override def eval(sketch: ItemsSketch[Any]): Any = {
+    val topK = right.eval().asInstanceOf[Int]
+    val items = sketch.getFrequentItems(ErrorType.NO_FALSE_POSITIVES)
+    val resultLength = math.min(items.length, topK)
+    val result = new Array[AnyRef](resultLength)
+    for (i <- 0 until resultLength) {
+      val row = items(i)
+      left.dataType match {
+        case _: LongType | _: DoubleType | _: IntegerType =>
+          result(i) = InternalRow.apply(row.getItem, row.getEstimate)
+        case _: StringType =>
+          val item = UTF8String.fromString(row.getItem.asInstanceOf[String])
+          result(i) = InternalRow.apply(item, row.getEstimate)
+        case dataType => throw new SparkUnsupportedOperationException(
```

let's extend `ExpectsInputTypes` and fail early on unexpected data types

---

Oh you already implemented it, let's remove this branch